geniza/corpus/models.py

```diff
@@ -65,3 +71,161 @@ def __str__(self):
+    # TODO footnotes for edition/translation
+    notes = models.TextField(blank=True)
+    old_input_by = models.CharField(
+        'Legacy input by', max_length=255,
+        help_text='Legacy input information from Google Sheets')
+    old_input_date = models.CharField(
+        'Legacy input date', max_length=255,
+        help_text='Legacy input date from Google Sheets')
+    created = models.DateTimeField(auto_now_add=True)
+    last_modified = models.DateTimeField(auto_now=True)
+
+    class Meta:
+        ordering = ['fragments__shelfmark']
+
+    def __str__(self):
+        return self.shelfmark
+
+    @property
+    def shelfmark(self):
+        '''shelfmarks for associated fragments'''
+        # TODO: honor order in text unit
+        return ' + '.join([block.fragment.shelfmark
+                           for block in self.textblock_set.all()])
+
+    @property
+    def collection(self):
+        '''collection (abbreviation) for associated fragments'''
+        # use set to ensure unique; sort for reliable output order
+        return ', '.join(sorted(set([block.fragment.collection.abbrev for
+                                block in self.textblock_set.all()
+                                if block.fragment.collection])))
+
+    def is_textblock(self):
+        '''Is this document part of a notated text block?'''
+        return any(bool(block.extent_label)
+                   for block in self.textblock_set.all())
+    is_textblock.short_description = 'Text Block?'
+    is_textblock.boolean = True
+
+    def all_languages(self):
+        return ','.join([str(lang) for lang in self.languages.all()])
+    all_languages.short_description = 'Language'
+
+    def tag_list(self):
+        return ", ".join(t.name for t in self.tags.all())
+    tag_list.short_description = 'tags'
+
+
+class TextBlock(models.Model):
+    '''The portion of a document that appears on a particular fragment.'''
+    document = models.ForeignKey(Document, on_delete=models.CASCADE)
+    fragment = models.ForeignKey(Fragment, on_delete=models.CASCADE)
+    RECTO = 'r'
+    VERSO = 'v'
+    RECTO_VERSO = 'rv'
+    RECTO_VERSO_CHOICES = [
+        (RECTO, 'recto'),
+        (VERSO, 'verso'),
+        (RECTO_VERSO, 'recto and verso'),
+    ]
+    side = models.CharField(blank=True, max_length=255,
+                            choices=RECTO_VERSO_CHOICES)
+    extent_label = models.CharField(blank=True, max_length=255)
+    order = models.PositiveIntegerField(
+        null=True, blank=True,
+        help_text='Order if there are multiple fragments. ' +
+                  'Top to bottom or right to left.')
+
+    class Meta:
+        ordering = ['order']
+
+    def __str__(self):
+        # combine shelfmark, side, and optionally text block
+        parts = [self.fragment.shelfmark, self.get_side_display(),
+                 self.extent_label]
+        return ' '.join(p for p in parts if p)
+
+    def thumbnail(self):
+        return self.fragment.iiif_thumbnails()
```

Comment:
confused about the function/naming of this - would it only ever return one thumbnail? seems like the method it delegates to could return multiple

Reply:
good point, that was inconsistent of me — I have some thoughts on what this might look like later when we can identify which image from a manifest is referenced, but probably best to name it simply & accurately for now!